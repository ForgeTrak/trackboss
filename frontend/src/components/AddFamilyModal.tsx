import React from 'react';
import {
    Modal,
    ModalOverlay,
    ModalContent,
    Button,
    Text,
    ModalFooter,
    Divider,
} from '@chakra-ui/react';

interface modalProps {
  isOpen: boolean,
  onClose: () => void,
}

export default function AddFamilyModal(props: modalProps) {
    return (
        <Modal isCentered size="xl" isOpen={props.isOpen} onClose={props.onClose}>
            <ModalOverlay />
            <ModalContent>
                <Text fontSize="2xl" pl={2} pr={2} textAlign="center">
                    Please contact a club board member to add members to your membership.
                </Text>
                <Divider />
                <ModalFooter>
                    <Button
                        color="green"
                        variant="ghost"
                        mr={3}
                        size="lg"
                        onClick={
                            () => {
                                props.onClose();
                            }
                        }
                    >
                        Accept
                    </Button>
                </ModalFooter>
            </ModalContent>
        </Modal>
    );
}
