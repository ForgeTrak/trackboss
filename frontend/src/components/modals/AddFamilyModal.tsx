import React, { useState } from 'react';
import {
    Modal,
    ModalOverlay,
    ModalContent,
    Button,
    Heading,
    ModalFooter,
    Divider,
    Box,
    Input,
    Spinner,
    Text,
} from '@chakra-ui/react';
import DatePicker from 'react-date-picker';
import moment from 'moment';
import WrappedSwitchInput from '../input/WrappedSwitchInput';
import { Member, PostNewMemberRequest } from '../../../../src/typedefs/member';

import 'react-date-picker/dist/DatePicker.css';
import '../../css/date-picker.css';
import { createMember } from '../../controller/member';

interface modalProps {
  isOpen: boolean,
  membershipAdmin: Member | undefined,
  token: string,
  onClose: () => void,
  refreshList: () => void,
}

export default function AddFamilyModal(props: modalProps) {
    const [firstName, setFirstName] = useState<string>();
    const [lastName, setLastName] = useState<string>();
    const [email, setEmail] = useState<string>();
    const [allowOnlineAccess, setAllowOnlineAccess] = useState<boolean>(true);
    const [birthDate, setBirthDate] = useState<Date>();
    const [isSaving, setIsSaving] = useState<boolean>(false);

    return (
        <Modal isCentered size="xl" isOpen={props.isOpen} onClose={props.onClose}>
            <ModalOverlay />
            <ModalContent>
                <Heading pl={2} pr={2} textAlign="center">
                    Add family member
                </Heading>
                <Divider />
                <Box width={300}>
                    <Box m={3}>
                        <Input
                            placeholder="First Name"
                            value={firstName}
                            onChange={
                                (e) => {
                                    setFirstName(e.target.value);
                                }
                            }
                        />
                    </Box>
                    <Box m={3}>
                        <Input
                            placeholder="Last Name"
                            value={lastName}
                            onChange={
                                (e) => {
                                    setLastName(e.target.value);
                                }
                            }
                        />
                    </Box>
                    <Box m={3}>
                        <Text>Birth date</Text>
                        <DatePicker
                            minDate={moment().subtract(100, 'years').toDate()}
                            maxDate={moment().subtract(1, 'months').toDate()}
                            onChange={
                                (date: Date) => {
                                    setBirthDate(date);
                                }
                            }
                            onCalendarClose={() => setBirthDate(birthDate)}
                            value={birthDate}
                        />
                    </Box>
                    <Box m={3}>
                        <Input
                            placeholder="email (if applicable)"
                            value={email}
                            onChange={
                                (e) => {
                                    setEmail(e.target.value);
                                }
                            }
                        />
                    </Box>
                    <Box m={3}>
                        <WrappedSwitchInput
                            maxWidth={300}
                            wrapperText="Allow online access?"
                            defaultChecked={allowOnlineAccess}
                            onSwitchChange={setAllowOnlineAccess}
                        />
                    </Box>
                </Box>
                <ModalFooter>
                    <Button
                        color="black"
                        variant="ghost"
                        hidden={isSaving}
                        mr={3}
                        size="lg"
                        onClick={
                            () => {
                                props.onClose();
                            }
                        }
                    >
                        Cancel
                    </Button>
                    <Button
                        color="green"
                        variant="ghost"
                        mr={3}
                        size="lg"
                        disabled={
                            (!firstName) ||
                            (!lastName) ||
                            (!birthDate) ||
                            (allowOnlineAccess && !email)
                        }
                        hidden={isSaving}
                        onClick={
                            async () => {
                                const familyMemberAdd : PostNewMemberRequest = {
                                    membershipId: props.membershipAdmin?.membershipId,
                                    // magic numberism - this is 'Member'.  anyone created this way will be a member
                                    // created by a membership admin.
                                    memberTypeId: 9,
                                    firstName,
                                    lastName,
                                    phoneNumber: props.membershipAdmin?.phoneNumber,
                                    occupation: '',
                                    email: '',
                                    birthdate: birthDate?.toLocaleDateString('en-CA'),
                                    dateJoined: props.membershipAdmin?.dateJoined,
                                    modifiedBy: props.membershipAdmin?.memberId || 0,
                                };
                                if (allowOnlineAccess) {
                                    familyMemberAdd.email = email;
                                }
                                setIsSaving(true);
                                await createMember(props.token, familyMemberAdd);
                                setIsSaving(false);
                                await props.refreshList();
                                props.onClose();
                            }
                        }
                    >
                        Save
                    </Button>
                    <Text hidden={!isSaving}>{`Saving ${firstName}, please wait`}</Text>
                    <Spinner hidden={!isSaving} color="orange" />
                </ModalFooter>
            </ModalContent>
        </Modal>
    );
}
