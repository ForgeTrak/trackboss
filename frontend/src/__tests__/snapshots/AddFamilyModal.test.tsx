import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import AddFamilyModal from '../../components/AddFamilyModal';

function onClose() {
    // void function
}

// Necessary to mock the chakra-ui modal portal, so that the renderer can target the modal instead of the page
const divWithChildrenMock = (children: any, identifier: any) => <div data-testId={identifier}>{children}</div>;
jest.mock('@chakra-ui/react', () => (
    {
        ...jest.requireActual('@chakra-ui/react'),
        PortalManager: jest.fn(({ children }) => divWithChildrenMock(children, 'portal')),
    }
));

describe('add member modal', () => {
    it('renders correctly', () => {
        const header = renderer.create(
            <BrowserRouter>
                <AddFamilyModal isOpen onClose={onClose} />
            </BrowserRouter>,
        ).toJSON();
        expect(header).toMatchSnapshot();
    });
});
