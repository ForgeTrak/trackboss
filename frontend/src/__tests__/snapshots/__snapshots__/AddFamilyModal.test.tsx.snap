// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add member modal renders correctly 1`] = `<span />`;
